File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/client/ClientConfigManifestLoader.java
package org.hyperledger.iroha.android.client;

import java.io.IOException;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Base64;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.function.BiConsumer;
import org.hyperledger.iroha.android.telemetry.TelemetryOptions;

/**
 * Loads {@link ClientConfig} instances from JSON manifests derived from {@code iroha_config}.
 *
 * <p>The loader keeps the manifest structure available via {@link ManifestContext} so callers can
 * inspect the inputs alongside the built {@link ClientConfig}.
 *
 * <p>Genesis {@code confidential_features} and {@code zk_policy_hash} values are consensus
 * fingerprints, not client proof policy. They are rejected at any depth so a customizer cannot
 * accidentally treat them as authoritative; proof capabilities must come from committed Torii
 * state and the on-chain verifying-key registry.
 */
public final class ClientConfigManifestLoader {

  private static final List<String> NON_AUTHORITATIVE_PRIVACY_POLICY_FIELDS =
      Collections.unmodifiableList(
          Arrays.asList(
              "confidential_features", "confidentialFeatures", "zk_policy_hash", "zkPolicyHash"));

  private ClientConfigManifestLoader() {}

  /** Loads the manifest at {@code manifestPath} using the default configuration. */
  public static LoadedClientConfig load(final Path manifestPath) throws IOException {
    return load(manifestPath, null);
  }

  /**
    * Loads the manifest at {@code manifestPath} and allows callers to customise the builder before
    * it is materialised.
    *
    * @param customizer invoked after the manifest is applied to the builder
    */
  public static LoadedClientConfig load(
      final Path manifestPath, final Customizer customizer) throws IOException {
    Objects.requireNonNull(manifestPath, "manifestPath");
    final byte[] payload =
        Files.readAllBytes(manifestPath);
    final String digest = sha256Hex(payload);
    return parse(manifestPath, payload, digest, customizer);
  }

  static LoadedClientConfig parse(
      final Path manifestPath,
      final byte[] payload,
      final String digest,
      final Customizer customizer) {
    final Map<String, Object> root = parseRoot(payload);
    rejectNonAuthoritativePrivacyPolicyFields(root, "manifest");
    final ClientConfig.Builder builder = ClientConfig.builder();
    final String chainId = optionalString(root.get("chain_id"), "chain_id");
    if (chainId != null) {
      if (chainId.isEmpty()) {
        throw new IllegalStateException("chain_id must be a non-empty string");
      }
      builder.setLocalSigningContext(new LocalSigningContext(chainId));
    }
    applyTorii(manifestPath, builder, root);
    applyRetry(builder, root);
    applyPendingQueue(manifestPath, builder, root);
    applyTelemetry(builder, root);
    final ManifestContext context =
        new ManifestContext(manifestPath, digest, immutableCopy(root));
    if (customizer != null) {
      customizer.accept(builder, context);
    }
    final ClientConfig config = builder.build();
    return new LoadedClientConfig(config, context, Instant.now());
  }

  private static Map<String, Object> parseRoot(final byte[] payload) {
    final String json = new String(payload, StandardCharsets.UTF_8).trim();
    if (json.isEmpty()) {
      throw new IllegalStateException("Client config manifest is empty");
    }
    final Object root = JsonParser.parse(json);
    return expectObject(root, "manifest");
  }

  private static void applyTorii(
      final Path manifestPath,
      final ClientConfig.Builder builder,
      final Map<String, Object> root) {
    final Map<String, Object> torii = expectObject(root.get("torii"), "torii");
    final String baseUri = requireString(torii, "base_uri");
    builder.setBaseUri(parseUri(baseUri, "torii.base_uri"));
    final String gateway =
        optionalString(torii.get("sorafs_gateway_uri"), "torii.sorafs_gateway_uri");
    if (gateway != null) {
      if (gateway.isEmpty()) {
        throw new IllegalStateException("torii.sorafs_gateway_uri must be a non-empty string");
      }
      builder.setSorafsGatewayUri(parseUri(gateway, "torii.sorafs_gateway_uri"));
    }
    final Long timeoutMs = optionalLong(torii.get("timeout_ms"), "torii.timeout_ms");
    if (timeoutMs != null) {
      if (timeoutMs < 0) {
        throw new IllegalStateException("torii.timeout_ms must be >= 0");
      }
      builder.setRequestTimeout(Duration.ofMillis(timeoutMs));
    }
    final Map<String, Object> headers =
        optionalObject(torii.get("default_headers"), "torii.default_headers");
    if (headers != null) {
      for (final Map.Entry<String, Object> entry : headers.entrySet()) {
        final String name = Objects.requireNonNull(entry.getKey(), "header name");
        final String value =
            requireStringValue(entry.getValue(), "torii.default_headers." + name);
        builder.putDefaultHeader(name, value);
      }
    }
  }

  private static void applyRetry(final ClientConfig.Builder builder, final Map<String, Object> root) {
    final Map<String, Object> retry = optionalObject(root.get("retry"), "retry");
    if (retry == null) {
      return;
    }
    final RetryPolicy.Builder policy = RetryPolicy.builder();
    final Integer attempts = optionalInt(retry.get("max_attempts"), "retry.max_attempts");
    if (attempts != null) {
      if (attempts < 1) {
        throw new IllegalStateException("retry.max_attempts must be >= 1");
      }
      policy.setMaxAttempts(attempts);
    }
    final Long baseDelayMs = optionalLong(retry.get("base_delay_ms"), "retry.base_delay_ms");
    if (baseDelayMs != null) {
      if (baseDelayMs < 0) {
        throw new IllegalStateException("retry.base_delay_ms must be >= 0");
      }
      policy.setBaseDelay(Duration.ofMillis(baseDelayMs));
    }
    final Long maxDelayMs = optionalLong(retry.get("max_delay_ms"), "retry.max_delay_ms");
    if (maxDelayMs != null) {
      if (maxDelayMs < 0) {
        throw new IllegalStateException("retry.max_delay_ms must be >= 0");
      }
      policy.setMaxDelay(Duration.ofMillis(maxDelayMs));
    }
    final Boolean retryServer =
        optionalBoolean(retry.get("retry_on_server_error"), "retry.retry_on_server_error");
    if (retryServer != null) {
      policy.setRetryOnServerError(retryServer);
    }
    final Boolean retryTooMany =
        optionalBoolean(
            retry.get("retry_on_too_many_requests"), "retry.retry_on_too_many_requests");
    if (retryTooMany != null) {
      policy.setRetryOnTooManyRequests(retryTooMany);
    }
    final Boolean retryNetwork =
        optionalBoolean(retry.get("retry_on_network_error"), "retry.retry_on_network_error");
    if (retryNetwork != null) {
      policy.setRetryOnNetworkError(retryNetwork);
    }
    final List<Object> codes = optionalArray(retry.get("retry_status_codes"), "retry.retry_status_codes");
    if (codes != null) {
      for (int index = 0; index < codes.size(); index++) {
        final Integer status =
            optionalInt(codes.get(index), "retry.retry_status_codes[" + index + "]");
        if (status == null) {
          throw new IllegalStateException(
              "retry.retry_status_codes[" + index + "] must be an integer");
        }
        policy.addRetryStatusCode(status);
      }
    }
    builder.setRetryPolicy(policy.build());
  }

  private static void applyPendingQueue(
      final Path manifestPath,
      final ClientConfig.Builder builder,
      final Map<String, Object> root) {
    final Map<String, Object> pending =
        optionalObject(root.get("pending_queue"), "pending_queue");
    if (pending == null) {
      return;
    }
    final String kind = optionalString(pending.get("kind"), "pending_queue.kind");
    if (kind == null || "memory".equalsIgnoreCase(kind)) {
      builder.setPendingQueue(null);
      return;
    }
    if ("file".equalsIgnoreCase(kind)) {
      final String pathRaw = requireString(pending, "path");
      final Path resolved = resolveRelative(manifestPath, pathRaw);
      builder.enableFilePendingQueue(resolved);
      return;
    }
    throw new IllegalStateException("Unsupported pending queue kind: " + kind);
  }

  private static void applyTelemetry(
      final ClientConfig.Builder builder, final Map<String, Object> root) {
    final Map<String, Object> telemetry =
        optionalObject(root.get("telemetry"), "telemetry");
    if (telemetry == null) {
      builder.setTelemetryOptions(TelemetryOptions.disabled());
      return;
    }
    final Boolean enabledFlag = optionalBoolean(telemetry.get("enabled"), "telemetry.enabled");
    final boolean enabled = enabledFlag == null || enabledFlag;
    final TelemetryOptions.Builder telemetryBuilder = TelemetryOptions.builder().setEnabled(enabled);
    if (enabled) {
      final Map<String, Object> redaction =
          optionalObject(telemetry.get("redaction"), "telemetry.redaction");
      if (redaction == null) {
        throw new IllegalStateException(
            "telemetry.redaction block required when telemetry.enabled is true");
      }
      telemetryBuilder.setTelemetryRedaction(parseRedaction(redaction));
    } else {
      telemetryBuilder.setTelemetryRedaction(TelemetryOptions.Redaction.disabled());
    }
    builder.setTelemetryOptions(telemetryBuilder.build());
    final String exporterName =
        optionalString(telemetry.get("exporter_name"), "telemetry.exporter_name");
    if (exporterName != null) {
      builder.setTelemetryExporterName(exporterName);
    }
  }

  private static TelemetryOptions.Redaction parseRedaction(final Map<String, Object> redaction) {
    final TelemetryOptions.Redaction.Builder builder = TelemetryOptions.Redaction.builder();
    final String saltB64 =
        optionalString(redaction.get("salt_b64"), "telemetry.redaction.salt_b64");
    final String saltHex =
        optionalString(redaction.get("salt_hex"), "telemetry.redaction.salt_hex");
    if (saltB64 != null && !saltB64.isEmpty()) {
      builder.setSalt(Base64.getDecoder().decode(saltB64));
    } else if (saltHex != null && !saltHex.trim().isEmpty()) {
      builder.setSaltHex(saltHex.trim());
    }
    final String saltVersion = requireString(redaction, "salt_version");
    builder.setSaltVersion(saltVersion);
    final String rotation =
        optionalString(redaction.get("rotation_id"), "telemetry.redaction.rotation_id");
    if (rotation != null && !rotation.isEmpty()) {
      builder.setRotationId(rotation);
    } else {
      builder.setRotationId(saltVersion);
    }
    final String algorithm =
        optionalString(redaction.get("algorithm"), "telemetry.redaction.algorithm");
    if (algorithm != null && !algorithm.isEmpty()) {
      builder.setAlgorithm(algorithm);
    }
    return builder.build();
  }

  private static Map<String, Object> immutableCopy(final Map<String, Object> input) {
    final Map<String, Object> copy = new LinkedHashMap<>();
    for (final Map.Entry<String, Object> entry : input.entrySet()) {
      copy.put(entry.getKey(), immutableValue(entry.getValue()));
    }
    return Collections.unmodifiableMap(copy);
  }

  private static Object immutableValue(final Object value) {
    if (value instanceof Map<?, ?> map) {
      final Map<String, Object> copy = new LinkedHashMap<>();
      for (final Map.Entry<?, ?> entry : map.entrySet()) {
        final Object key = entry.getKey();
        if (!(key instanceof String)) {
          continue;
        }
        copy.put((String) key, immutableValue(entry.getValue()));
      }
      return Collections.unmodifiableMap(copy);
    }
    if (value instanceof List<?> list) {
      final List<Object> copy = new ArrayList<>(list.size());
      for (final Object element : list) {
        copy.add(immutableValue(element));
      }
      return Collections.unmodifiableList(copy);
    }
    return value;
  }

  private static void rejectNonAuthoritativePrivacyPolicyFields(
      final Object value, final String path) {
    if (value instanceof Map<?, ?> map) {
      for (final Map.Entry<?, ?> entry : map.entrySet()) {
        if (!(entry.getKey() instanceof String key)) {
          continue;
        }
        final String fieldPath = path + "." + key;
        if (NON_AUTHORITATIVE_PRIVACY_POLICY_FIELDS.contains(key)) {
          throw new IllegalStateException(
              fieldPath
                  + " is not authoritative client proof policy; use committed "
                  + "/v1/privacy/capabilities and the on-chain verifying-key registry");
        }
        rejectNonAuthoritativePrivacyPolicyFields(entry.getValue(), fieldPath);
      }
    } else if (value instanceof List<?> list) {
      for (int index = 0; index < list.size(); index++) {
        rejectNonAuthoritativePrivacyPolicyFields(list.get(index), path + "[" + index + "]");
      }
    }
  }

  private static Map<String, Object> expectObject(final Object value, final String path) {
    if (value == null) {
      throw new IllegalStateException(path + " block is required");
    }
    if (!(value instanceof Map<?, ?> map)) {
      throw new IllegalStateException(path + " is not a JSON object");
    }
    final Map<String, Object> copy = new LinkedHashMap<>();
    for (final Map.Entry<?, ?> entry : map.entrySet()) {
      final Object key = entry.getKey();
      if (!(key instanceof String)) {
        continue;
      }
      copy.put((String) key, entry.getValue());
    }
    return copy;
  }

  private static Map<String, Object> optionalObject(final Object value, final String path) {
    if (value == null) {
      return null;
    }
    if (!(value instanceof Map<?, ?> map)) {
      throw new IllegalStateException(path + " is not a JSON object");
    }
    final Map<String, Object> copy = new LinkedHashMap<>();
    for (final Map.Entry<?, ?> entry : map.entrySet()) {
      final Object key = entry.getKey();
      if (key instanceof String str) {
        copy.put(str, entry.getValue());
      }
    }
    return copy;
  }

  private static List<Object> optionalArray(final Object value, final String path) {
    if (value == null) {
      return null;
    }
    if (!(value instanceof List<?> list)) {
      throw new IllegalStateException(path + " is not an array");
    }
    return new ArrayList<>(list);
  }

  private static String requireString(final Map<String, Object> object, final String key) {
    final Object value = object.get(key);
    if (value == null) {
      throw new IllegalStateException("Missing required field: " + key);
    }
    final String normalized = requireStringValue(value, key);
    if (normalized.isEmpty()) {
      throw new IllegalStateException("Field " + key + " must be a non-empty string");
    }
    return normalized;
  }

  private static String requireStringValue(final Object value, final String path) {
    if (!(value instanceof String string)) {
      throw new IllegalStateException(path + " must be a string");
    }
    return string;
  }

  private static String optionalString(final Object value, final String path) {
    if (value == null) {
      return null;
    }
    return requireStringValue(value, path);
  }

  private static Integer optionalInt(final Object value, final String path) {
    if (value == null) {
      return null;
    }
    if (value instanceof Number number) {
      final long longValue = JsonNumbers.asLong(number, path);
      if (longValue < Integer.MIN_VALUE || longValue > Integer.MAX_VALUE) {
        throw new IllegalStateException("Integer value out of range: " + value);
      }
      return (int) longValue;
    }
    if (!(value instanceof String normalized)) {
      throw new IllegalStateException(path + " must be an integer or decimal integer string");
    }
    try {
      return Integer.parseInt(normalized);
    } catch (final NumberFormatException ex) {
      throw new IllegalStateException(
          path + " must be an integer or decimal integer string", ex);
    }
  }

  private static Long optionalLong(final Object value, final String path) {
    if (value == null) {
      return null;
    }
    if (value instanceof Number number) {
      return JsonNumbers.asLong(number, path);
    }
    if (!(value instanceof String normalized)) {
      throw new IllegalStateException(path + " must be an integer or decimal integer string");
    }
    try {
      return Long.parseLong(normalized);
    } catch (final NumberFormatException ex) {
      throw new IllegalStateException(
          path + " must be an integer or decimal integer string", ex);
    }
  }

  private static Boolean optionalBoolean(final Object value, final String path) {
    if (value == null) {
      return null;
    }
    if (value instanceof Boolean bool) {
      return bool;
    }
    if (!(value instanceof String normalized)) {
      throw new IllegalStateException(path + " must be a boolean or boolean string");
    }
    final String lower = normalized.trim().toLowerCase(Locale.ROOT);
    if ("true".equals(lower) || "1".equals(lower) || "yes".equals(lower)) {
      return true;
    }
    if ("false".equals(lower) || "0".equals(lower) || "no".equals(lower)) {
      return false;
    }
    throw new IllegalStateException("Unsupported boolean value for " + path + ": " + value);
  }

  private static URI parseUri(final String value, final String field) {
    try {
      return URI.create(value);
    } catch (final IllegalArgumentException ex) {
      throw new IllegalStateException("Invalid URI for " + field + ": " + value, ex);
    }
  }

  private static Path resolveRelative(final Path manifestPath, final String target) {
    final Path baseDir = manifestPath.toAbsolutePath().getParent();
    if (baseDir == null) {
      return Path.of(target).toAbsolutePath();
    }
    return baseDir.resolve(target).normalize();
  }

  private static byte[] hexToBytes(final String hex) {
    final String normalized = hex.trim();
    if ((normalized.length() & 1) == 1) {
      throw new IllegalStateException("Hex value must have an even number of characters");
    }
    final byte[] result = new byte[normalized.length() / 2];
    for (int i = 0; i < normalized.length(); i += 2) {
      final int hi = Character.digit(normalized.charAt(i), 16);
      final int lo = Character.digit(normalized.charAt(i + 1), 16);
      if (hi < 0 || lo < 0) {
        throw new IllegalStateException("Hex value contains invalid characters");
      }
      result[i / 2] = (byte) ((hi << 4) | lo);
    }
    return result;
  }

  static String sha256Hex(final byte[] payload) {
    final MessageDigest digest;
    try {
      digest = MessageDigest.getInstance("SHA-256");
    } catch (final NoSuchAlgorithmException ex) {
      throw new IllegalStateException("SHA-256 digest not available", ex);
    }
    final byte[] hash = digest.digest(payload);
    final StringBuilder builder = new StringBuilder(hash.length * 2);
    for (final byte value : hash) {
      builder.append(String.format(Locale.ROOT, "%02x", value));
    }
    return builder.toString();
  }

  /** Allows callers to customise the builder after the manifest has been applied. */
  @FunctionalInterface
  public interface Customizer extends BiConsumer<ClientConfig.Builder, ManifestContext> {
    @Override
    void accept(ClientConfig.Builder builder, ManifestContext context);
  }

  /** Immutable manifest metadata passed to customisers and returned in {@link LoadedClientConfig}. */
  public static final class ManifestContext {
    private final Path manifestPath;
    private final String digest;
    private final Map<String, Object> manifest;

    ManifestContext(
        final Path manifestPath, final String digest, final Map<String, Object> manifest) {
      this.manifestPath = manifestPath;
      this.digest = digest;
      this.manifest = manifest;
    }

    public Path manifestPath() {
      return manifestPath;
    }

    public String digest() {
      return digest;
    }

    public Map<String, Object> manifest() {
      return manifest;
    }
  }

  /** Wraps the built {@link ClientConfig} together with provenance metadata. */
  public static final class LoadedClientConfig {
    private final ClientConfig clientConfig;
    private final ManifestContext context;
    private final Instant loadedAt;

    LoadedClientConfig(
        final ClientConfig clientConfig,
        final ManifestContext context,
        final Instant loadedAt) {
      this.clientConfig = Objects.requireNonNull(clientConfig, "clientConfig");
      this.context = Objects.requireNonNull(context, "context");
      this.loadedAt = Objects.requireNonNull(loadedAt, "loadedAt");
    }

    public ClientConfig clientConfig() {
      return clientConfig;
    }

    public ManifestContext context() {
      return context;
    }

    public Instant loadedAt() {
      return loadedAt;
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/client/ClientConfigManifestLoader.kt
package org.hyperledger.iroha.sdk.client

import java.io.IOException
import java.net.URI
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.security.MessageDigest
import java.time.Duration
import java.time.Instant
import java.util.Collections
import java.util.LinkedHashMap
import java.util.function.BiConsumer
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi
import org.hyperledger.iroha.sdk.telemetry.Redaction
import org.hyperledger.iroha.sdk.telemetry.TelemetryOptions

/**
 * Loads [ClientConfig] instances from JSON manifests derived from `iroha_config`.
 *
 * The loader keeps the manifest structure available via [ManifestContext] so callers can
 * inspect the inputs alongside the built [ClientConfig].
 */
@OptIn(ExperimentalEncodingApi::class)
object ClientConfigManifestLoader {

    @JvmStatic @Throws(IOException::class) fun load(manifestPath: Path): LoadedClientConfig = load(manifestPath, null)

    @JvmStatic @Throws(IOException::class)
    fun load(manifestPath: Path, customizer: Customizer?): LoadedClientConfig {
        val payload = Files.readAllBytes(manifestPath)
        val digest = sha256Hex(payload)
        return parse(manifestPath, payload, digest, customizer)
    }

    @JvmStatic
    internal fun parse(manifestPath: Path, payload: ByteArray, digest: String, customizer: Customizer?): LoadedClientConfig {
        val root = parseRoot(payload)
        val builder = ClientConfig.builder()
        applyTorii(manifestPath, builder, root)
        applyRetry(builder, root)
        applyPendingQueue(manifestPath, builder, root)
        applyTelemetry(builder, root)
        val context = ManifestContext(manifestPath, digest, immutableCopy(root))
        customizer?.accept(builder, context)
        return LoadedClientConfig(builder.build(), context, Instant.now())
    }

    private fun parseRoot(payload: ByteArray): MutableMap<String, Any?> {
        val json = String(payload, StandardCharsets.UTF_8).trim()
        check(json.isNotEmpty()) { "Client config manifest is empty" }
        return expectObject(JsonParser.parse(json), "manifest")
    }

    private fun applyTorii(manifestPath: Path, builder: ClientConfig.Builder, root: Map<String, Any?>) {
        val torii = expectObject(root["torii"], "torii")
        builder.setBaseUri(parseUri(requireString(torii, "base_uri"), "torii.base_uri"))
        val gateway = optionalString(torii["sorafs_gateway_uri"], "torii.sorafs_gateway_uri")
        if (gateway != null) {
            check(gateway.isNotEmpty()) { "torii.sorafs_gateway_uri must be a non-empty string" }
            builder.setSorafsGatewayUri(parseUri(gateway, "torii.sorafs_gateway_uri"))
        }
        val timeoutMs = optionalLong(torii["timeout_ms"], "torii.timeout_ms")
        if (timeoutMs != null) {
            check(timeoutMs >= 0) { "torii.timeout_ms must be >= 0" }
            builder.setRequestTimeout(Duration.ofMillis(timeoutMs))
        }
        val headers = optionalObject(torii["default_headers"], "torii.default_headers")
        headers?.forEach { (name, value) ->
            builder.putDefaultHeader(name, requireStringValue(value, "torii.default_headers.$name"))
        }
    }

    private fun applyRetry(builder: ClientConfig.Builder, root: Map<String, Any?>) {
        val retry = optionalObject(root["retry"], "retry") ?: return
        val policy = RetryPolicy.builder()
        optionalInt(retry["max_attempts"], "retry.max_attempts")?.let {
            check(it >= 1) { "retry.max_attempts must be >= 1" }
            policy.setMaxAttempts(it)
        }
        optionalLong(retry["base_delay_ms"], "retry.base_delay_ms")?.let {
            check(it >= 0) { "retry.base_delay_ms must be >= 0" }
            policy.setBaseDelay(Duration.ofMillis(it))
        }
        optionalLong(retry["max_delay_ms"], "retry.max_delay_ms")?.let {
            check(it >= 0) { "retry.max_delay_ms must be >= 0" }
            policy.setMaxDelay(Duration.ofMillis(it))
        }
        optionalBoolean(retry["retry_on_server_error"], "retry.retry_on_server_error")?.let { policy.setRetryOnServerError(it) }
        optionalBoolean(retry["retry_on_too_many_requests"], "retry.retry_on_too_many_requests")?.let { policy.setRetryOnTooManyRequests(it) }
        optionalBoolean(retry["retry_on_network_error"], "retry.retry_on_network_error")?.let { policy.setRetryOnNetworkError(it) }
        optionalArray(retry["retry_status_codes"], "retry.retry_status_codes")?.forEachIndexed { index, value ->
            val status = optionalInt(value, "retry.retry_status_codes[$index]")
            check(status != null) { "retry.retry_status_codes[$index] must be an integer" }
            policy.addRetryStatusCode(status)
        }
        builder.setRetryPolicy(policy.build())
    }

    private fun applyPendingQueue(manifestPath: Path, builder: ClientConfig.Builder, root: Map<String, Any?>) {
        val pending = optionalObject(root["pending_queue"], "pending_queue") ?: return
        val kind = optionalString(pending["kind"], "pending_queue.kind")
        if (kind == null || "memory".equals(kind, ignoreCase = true)) { builder.setPendingQueue(null); return }
        if ("file".equals(kind, ignoreCase = true)) {
            val pathRaw = requireString(pending, "path")
            builder.enableFilePendingQueue(resolveRelative(manifestPath, pathRaw)); return
        }
        throw IllegalStateException("Unsupported pending queue kind: $kind")
    }

    private fun applyTelemetry(builder: ClientConfig.Builder, root: Map<String, Any?>) {
        val telemetry = optionalObject(root["telemetry"], "telemetry")
        if (telemetry == null) { builder.setTelemetryOptions(TelemetryOptions.disabled()); return }
        val enabledFlag = optionalBoolean(telemetry["enabled"], "telemetry.enabled")
        val enabled = enabledFlag == null || enabledFlag
        val telemetryBuilder = TelemetryOptions.builder().setEnabled(enabled)
        if (enabled) {
            val redaction = optionalObject(telemetry["redaction"], "telemetry.redaction")
            check(redaction != null) { "telemetry.redaction block required when telemetry.enabled is true" }
            telemetryBuilder.setTelemetryRedaction(parseRedaction(redaction))
        } else { telemetryBuilder.setTelemetryRedaction(Redaction.disabled()) }
        builder.setTelemetryOptions(telemetryBuilder.build())
        optionalString(telemetry["exporter_name"], "telemetry.exporter_name")?.let { builder.setTelemetryExporterName(it) }
    }

    private fun parseRedaction(redaction: Map<String, Any?>): Redaction {
        val rb = Redaction.builder()
        val saltB64 = optionalString(redaction["salt_b64"], "telemetry.redaction.salt_b64")
        val saltHex = optionalString(redaction["salt_hex"], "telemetry.redaction.salt_hex")
        if (!saltB64.isNullOrEmpty()) rb.setSalt(Base64.decode(saltB64))
        else if (!saltHex.isNullOrBlank()) rb.setSaltHex(saltHex.trim())
        rb.setSaltVersion(requireString(redaction, "salt_version"))
        val rotation = optionalString(redaction["rotation_id"], "telemetry.redaction.rotation_id")
        rb.setRotationId(if (!rotation.isNullOrEmpty()) rotation else requireString(redaction, "salt_version"))
        optionalString(redaction["algorithm"], "telemetry.redaction.algorithm")?.takeIf { it.isNotEmpty() }?.let { rb.setAlgorithm(it) }
        return rb.build()
    }

    private fun immutableCopy(input: Map<String, Any?>): Map<String, Any?> {
        val copy = LinkedHashMap<String, Any?>()
        for ((k, v) in input) copy[k] = immutableValue(v)
        return Collections.unmodifiableMap(copy)
    }

    private fun immutableValue(value: Any?): Any? = when (value) {
        is Map<*, *> -> Collections.unmodifiableMap(LinkedHashMap<String, Any?>().also { copy -> value.forEach { (k, v) -> if (k is String) copy[k] = immutableValue(v) } })
        is List<*> -> Collections.unmodifiableList(value.map { immutableValue(it) })
        else -> value
    }

    @Suppress("UNCHECKED_CAST")
    private fun expectObject(value: Any?, path: String): MutableMap<String, Any?> {
        check(value != null) { "$path block is required" }
        check(value is Map<*, *>) { "$path is not a JSON object" }
        val copy = LinkedHashMap<String, Any?>()
        for ((k, v) in value) { if (k is String) copy[k] = v }
        return copy
    }

    @Suppress("UNCHECKED_CAST")
    private fun optionalObject(value: Any?, path: String): MutableMap<String, Any?>? {
        if (value == null) return null
        check(value is Map<*, *>) { "$path is not a JSON object" }
        val copy = LinkedHashMap<String, Any?>()
        for ((k, v) in value) { if (k is String) copy[k] = v }
        return copy
    }

    private fun optionalArray(value: Any?, path: String): List<Any?>? {
        if (value == null) return null
        check(value is List<*>) { "$path is not an array" }
        return ArrayList(value)
    }

    private fun requireString(obj: Map<String, Any?>, key: String): String {
        val value = obj[key]
        check(value != null) { "Missing required field: $key" }
        val normalized = requireStringValue(value, key)
        check(normalized.isNotEmpty()) { "Field $key must be a non-empty string" }
        return normalized
    }

    private fun requireStringValue(value: Any?, path: String): String {
        check(value is String) { "$path must be a string" }
        return value
    }

    private fun optionalString(value: Any?, path: String): String? =
        if (value == null) null else requireStringValue(value, path)

    private fun optionalInt(value: Any?, path: String): Int? {
        if (value == null) return null
        if (value is Number) {
            return JsonNumbers.asInt(value, path)
        }
        check(value is String) { "$path must be an integer or decimal integer string" }
        return value.toIntOrNull()
            ?: throw IllegalStateException("$path must be an integer or decimal integer string")
    }

    private fun optionalLong(value: Any?, path: String): Long? {
        if (value == null) return null
        if (value is Number) return JsonNumbers.asLong(value, path)
        check(value is String) { "$path must be an integer or decimal integer string" }
        return value.toLongOrNull()
            ?: throw IllegalStateException("$path must be an integer or decimal integer string")
    }

    private fun optionalBoolean(value: Any?, path: String): Boolean? {
        if (value == null) return null
        if (value is Boolean) return value
        check(value is String) { "$path must be a boolean or boolean string" }
        return when (value.trim().lowercase()) {
            "true", "1", "yes" -> true
            "false", "0", "no" -> false
            else -> throw IllegalStateException("Unsupported boolean value for $path: $value")
        }
    }

    private fun parseUri(value: String, field: String): URI {
        try { return URI.create(value) } catch (ex: IllegalArgumentException) { throw IllegalStateException("Invalid URI for $field: $value", ex) }
    }

    private fun resolveRelative(manifestPath: Path, target: String): Path {
        val baseDir = manifestPath.toAbsolutePath().parent ?: return Paths.get(target).toAbsolutePath()
        return baseDir.resolve(target).normalize()
    }

    private fun hexToBytes(hex: String): ByteArray {
        val normalized = hex.trim()
        check(normalized.length % 2 == 0) { "Hex value must have an even number of characters" }
        val result = ByteArray(normalized.length / 2)
        for (i in normalized.indices step 2) {
            val hi = Character.digit(normalized[i], 16)
            val lo = Character.digit(normalized[i + 1], 16)
            check(hi >= 0 && lo >= 0) { "Hex value contains invalid characters" }
            result[i / 2] = ((hi shl 4) or lo).toByte()
        }
        return result
    }

    @JvmStatic
    internal fun sha256Hex(payload: ByteArray): String {
        val hash = MessageDigest.getInstance("SHA-256").digest(payload)
        return hash.joinToString("") { String.format("%02x", it) }
    }

    /** Allows callers to customise the builder after the manifest has been applied. */
    fun interface Customizer : BiConsumer<ClientConfig.Builder, ManifestContext>

    /** Immutable manifest metadata passed to customisers and returned in [LoadedClientConfig]. */
    class ManifestContext internal constructor(
        private val manifestPath: Path,
        private val digest: String,
        private val manifest: Map<String, Any?>
    ) {
        fun manifestPath(): Path = manifestPath
        fun digest(): String = digest
        fun manifest(): Map<String, Any?> = manifest
    }

    /** Wraps the built [ClientConfig] together with provenance metadata. */
    class LoadedClientConfig internal constructor(
        private val clientConfig: ClientConfig,
        private val context: ManifestContext,
        private val loadedAt: Instant
    ) {
        fun clientConfig(): ClientConfig = clientConfig
        fun context(): ManifestContext = context
        fun loadedAt(): Instant = loadedAt
    }
}
